examples/eks-cluster-with-external-dns/main.tf

```diff
@@ -0,0 +1,196 @@
+terraform {
+  required_version = ">= 1.0.1"
+
+  required_providers {
+    aws = {
+      source  = "hashicorp/aws"
+      version = ">= 3.66.0"
+    }
+    kubernetes = {
+      source  = "hashicorp/kubernetes"
+      version = ">= 2.6.1"
+    }
+    helm = {
+      source  = "hashicorp/helm"
+      version = ">= 2.4.1"
+    }
+  }
+
+  backend "local" {
+    path = "local_tf_state/terraform-main.tfstate"
+  }
+}
+
+data "aws_region" "current" {}
+
+data "aws_availability_zones" "available" {}
+
+data "aws_eks_cluster" "cluster" {
+  name = module.eks_cluster.eks_cluster_id
+}
+
+data "aws_eks_cluster_auth" "cluster" {
+  name = module.eks_cluster.eks_cluster_id
+}
+
+data "aws_acm_certificate" "issued" {
+  domain   = var.acm_certificate_domain
+  statuses = ["ISSUED"]
+}
+
+data "aws_route53_zone" "selected" {
+  name = var.eks_cluster_domain
+}
+
+provider "aws" {
+  region = "us-west-1"
+  alias  = "default"
+}
+
+provider "kubernetes" {
+  experiments {
+    manifest_resource = true
+  }
+  host                   = data.aws_eks_cluster.cluster.endpoint
+  cluster_ca_certificate = base64decode(data.aws_eks_cluster.cluster.certificate_authority.0.data)
+  token                  = data.aws_eks_cluster_auth.cluster.token
+}
+
+provider "helm" {
+  kubernetes {
+    host                   = data.aws_eks_cluster.cluster.endpoint
+    token                  = data.aws_eks_cluster_auth.cluster.token
+    cluster_ca_certificate = base64decode(data.aws_eks_cluster.cluster.certificate_authority.0.data)
+  }
+}
+
+locals {
+  tenant      = "aws01"   # AWS account name or unique id for tenant
+  environment = "preprod" # Environment area eg., preprod or prod
+  zone        = "dev"     # Environment with in one sub_tenant or business unit
+
+  kubernetes_version = "1.21"
+
+  vpc_cidr       = "10.0.0.0/16"
+  vpc_name       = join("-", [local.tenant, local.environment, local.zone, "vpc"])
+  eks_cluster_id = join("-", [local.tenant, local.environment, local.zone, "eks"])
+
+  terraform_version = "Terraform v1.0.1"
+}
+
+#---------------------------------------------------------------
+# VPC
+#---------------------------------------------------------------
+
+module "aws_vpc" {
+  source  = "terraform-aws-modules/vpc/aws"
+  version = "3.13.0"
+
+  name = local.vpc_name
+  cidr = local.vpc_cidr
+  azs  = data.aws_availability_zones.available.names
+
+  public_subnets  = [for k, v in data.aws_availability_zones.available.names : cidrsubnet(local.vpc_cidr, 8, k)]
```

Comment:
This is going to create subnets for all available AZs on the region, that may lead into issues where EKS cluster can't be deployed to that AZ(s) or other issues where one of the AZs is not supported (Very common in us-east-1 for example). We should either set it to something like:
```
slice(data.aws_availability_zones.available.names, 0, 3)
```
or in the data source, we can add filter:
```
data "aws_availability_zones" "available" {
  state = "available"
  # Specify AZs to avoid EKS cluster creation error due to reduced capacity in an AZ.
  # Change the AZ names per capacity available in the region you selected.
  # https://aws.amazon.com/premiumsupport/knowledge-center/eks-cluster-creation-errors/
  filter {
    name   = "zone-name"
    values = ["us-west-2a", "us-west-2b", "us-west-2c"]
  }
}
```

Reply:
Well, we need to do this across the board in all our examples. Lets not hold this addon for this. We can do follow up PR to add that.